unitarity constraint in the s->inf limit

wkotlarski opened the pull request:
I've decided to split implementation of unitarity constraints for $s \to \infty$ from #524 into a separate PR.

@Expander, how should we present results? Should we implement
```
Block TREELEVELUNITARITY
   0    1.00000000E+00 # Tree-level unitarity limits fulfilled or not
   1    7.32883464E+00 # Maximal scattering eigenvalue
```
like SARAH (see Sec. 4.2.3 in [here](https://arxiv.org/pdf/1805.07306.pdf))

TODO:
- [x] decide at which scale should unitarity be computed
- [x] `max_scattering_eigenvalue_infinite_s` function should return the entire matrix of eigenvalues for testing
- [x] write tests
- [x] shift indices to C++ convention at the level of expression
- [x] indent generated code
- [x] move `set_unitarity_infinite_s` from `templates/slha_io.cpp.in` to `src/slha_io.cpp`
- [x] mathlink interface

Validation:
- [x] SM
- [ ] SSM
- [ ] MSSM?

Expander commented:
> @Expander, how should we present results? Should we implement
> 
> ```
> Block TREELEVELUNITARITY
>    0    1.00000000E+00 # Tree-level unitarity limits fulfilled or not
>    1    7.32883464E+00 # Maximal scattering eigenvalue
> ```
> 
> like SARAH (see Sec. 4.2.3 in [here](https://arxiv.org/pdf/1805.07306.pdf))

This would have the benefit of being compatible with SARAH. However, it would have the disadvantage that a user cannot easily use FS and SARAH together, because the block may be overwritten by one of the programs.

An alternative form could be:
~~~
Block FlexibleSUSYUnitarity # or FlexibleUnitarity
   0    1.00000000E+00 # Tree-level unitarity limits fulfilled or not
   1    7.32883464E+00 # Maximal scattering tree-level eigenvalue
   2    1.00000000E+00 # 1-loop level unitarity limits fulfilled or not
   3    7.32883464E+00 # Maximal scattering 1-loop level eigenvalue
~~~
which would allow us to add further fields to the block (like 1-loop unitarity limits etc.). Would this make sense?

wkotlarski commented:
I've implemented it as follows
```
Block FlexibleSUSYUnitarity Q= 1.73340000E+02
     0     1.00000000E+00   # Tree-level unitarity limits fulfilled or not
     1     1.00121998E-02    # max(|re(a0)|)
```
Unitarity in the $s\to\infty$ limit is always computed at a fixed scale so I've also added it to the output.

wkotlarski commented:
Should we implement the `mathlink` part as a new handle, e.g. `FS<model>CalculateUnitarity` (and later also `FS<model>CalculateUnitaritywTrilinears`)?

Expander commented:
> Should we implement the `mathlink` part as a new handle, e.g. `FS<model>CalculateUnitarity` (and later also `FS<model>CalculateUnitaritywTrilinears`)?

Do you mean that these functions should return a *new* handle? Or do you mean to just have separate (new) functions?

wkotlarski commented:
I meant that since we don't implement unitarity as an observable we should not return it from `FS<model>CalculateObservables[handle]` but probably create a new function, e.g. `FS<model>CalculateUnitarity`.

Expander commented:
> I meant that since we don't implement unitarity as an observable we should not return it from `FS<model>CalculateObservables[handle]` but probably create a new function, e.g. `FS<model>CalculateUnitarity`.

Inventing a new function is a good idea, indeed. I'd prefer this.

Expander commented:
Excellent!

One remark, though: Can we rename `Scale` to something else, like `RenormalizationScale` or so? I find `Scale` to be too generic.

wkotlarski commented:
I'm confused how does the `FS<model>ToSLHA` function works. It doesn't create the `FlexibleSUSYUnitarity` block. But how does it know which blocks to create at all?

Expander commented on doc/observables.rst:
```diff
@@ -24,3 +24,4 @@ List of observables computed by ``FlexibleSUSY``
   - LO order (tree-level or loop-induced) decays of Higgs scalars into non-SM state
   - LO order (tree-level or loop-induced) decays of non-Higgs scalars
 
+- Unitarity costraints
```
costraints -> constraints

Expander commented on meta/FlexibleSUSY.m:
```diff
@@ -500,6 +503,20 @@ FlexibleSUSY model file (FlexibleSUSY.m).
               FlexibleSUSY`FSCalculateDecays = False
            ];
+
+           With[{sarahVersion = DecomposeVersionString[SA`Version],
```
It seems to me that the following `If` expression is a copy of the one in `CheckSARAHVersion[]`. Maybe both can be replaced by `VersionOrder[]` (or we move the If statement to a separate function)?

wkotlarski replied:
Where should we put it? In `Utils`?

Expander replied:
Yes, this would be a good place.

wkotlarski replied:
moved to #529

Expander commented on meta/FSMathLink.m:
```diff
@@ -245,6 +249,9 @@
     "virtual void " <> CreateSpectrumDecaysCalculationName[] <> "(const softsusy::QedQcd&, const Physical_input&, const FlexibleDecay_settings&) = 0;";
 
+CreateSpectrumUnitarityInterface[modelName_] :=
+    "virtual std::tuple<bool, double, double, Eigen::MatrixXcd> calculate_unitarity() = 0;";
```
I'd recommend to use a `struct` instead of a `tuple`, because a struct has "named" elements, which improves readability.

wkotlarski replied:
I agree. I was also thinking about it. This tuple got out of hand (it started with 2 elements). I'll make the change.

wkotlarski replied:
Fixed in a756dd999fad3f0bf06a9b5460987428e744e6cd.

Expander commented on templates/unitarity.cpp.in:
```diff
@@ -0,0 +1,76 @@
+
+static constexpr int size = @scatteringPairsLength@;
+
+inline
```
I'd recommend to put these three `inline` functions into an anonymous namespace, which expresses the fact that these functions are local to this translation unit.

wkotlarski replied:
Fixed in 366bca7cb9d71186a6c903a32f53318c997e2a56.

Expander commented on meta/FSMathLink.m:
```diff
@@ -245,6 +249,9 @@
 CreateSpectrumDecaysInterface[modelName_] :=
     "virtual void " <> CreateSpectrumDecaysCalculationName[] <> "(const softsusy::QedQcd&, const Physical_input&, const FlexibleDecay_settings&) = 0;";
 
+CreateSpectrumUnitarityInterface[modelName_] :=
```
I think this function is defined twice, see Line 271.

wkotlarski replied:
Removed in 8e291e7.

Expander commented:
One design question: It seems you are generating the C++ interface functions (for example `CreateMathLinkUnitarityCalculation` and others) by the meta code, depending on whether the user wants to calculate the unitarity constraints or not.

Initially I'd have done this slightly differently: One could always define the interface functions by moving them into the template files. However, one would fill their body by either a `throw` statement (if the user did not enable the unitarity constraints) or the actual calculation  (if the user did enable the unitarity constraints).

Your approach has the characteristic that one cannot call the C++ functions if the user has not enabled the unitarity constraint calculation, because they are simply not defined in the C++ code. However, in this approach you'd need to generate more stuff, i.e. the meta code becomes a bit larger and there is more entanglement between the meta code and the C++ code.

My naive approach would have the characteristic that one can always call the C++ functions; however, they may throw an exception if the user has not enabled the unitarity constraint calculation. OTOH, one would need to generate less C++ code and could move more code into the C++ template files. So there would be a bit less meta code and a bit less entanglement between the meta code and the C++ code.

What is your reason to implement this approach? (I'm not saying that this is bad, I'm just curious -- maybe you have a good reason. I guess it is simply a design choice with pros/cons.)

wkotlarski commented:
I've moved the stuff related to SLHA output out of the meta code in 08f4438, and the mathlink part in 5929ff6.

wkotlarski commented:
@Expander, I think we can also bring this one in. It seems to work fine and passes simple comparisons with SARAH. We can postpone detailed validation to FlexibleSUSY 3.0 manual. And for the time being, we just won't advertise it.

Expander commented on meta/Unitarity.m:
```diff
@@ -0,0 +1,139 @@
+
+FSScatteringPairsSizes[pairs_] := Map[TreeMasses`GetDimension, pairs, {2}];
+
+InfiniteS[a0Input_, generationSizes_, FSScatteringPairs_] := Module[{params = Parameters`FindAllParametersClassified[a0Input], paramsCPP, mixingCPP, a0, decrementIndices, removeFactor=16*Pi},
```
Just one comment to this function: The function body looks a bit convoluted. I think we need more testing to ensure it is correct, as you suggested.

Expander commented on src/unitarity.hpp:
```diff
@@ -0,0 +1,43 @@
+
+namespace flexiblesusy {
+
+struct UnitarityInfiniteS {
```
I'd suggest to default-initialize all four members of this struct to avoid problems.

wkotlarski replied:
Done in 8ff4b88.

Expander commented on templates/unitarity.cpp.in:
```diff
@@ -0,0 +1,80 @@
+namespace {
+
+inline
+double Sqrt2(int i, int j) {
```
Would it make sense to rename this function? Maybe to `Sqrt2IfEq` or so?

wkotlarski replied:
I could do that. But there's actually a logic behind it. That's how Mark called it. And this whole thing is just a wrapper around his calculation. But I don't have a strong opinion on this issue. So if you like I can rename it.

Expander replied:
Ah, good point. It would make sense to leave the name for comparison with SARAH. Then I'd propose to leave the name as it is.

Expander commented on test/test_CMSSM_unitarity.cpp:
```diff
@@ -0,0 +1,534 @@
+   auto unitarityMatrix = CMSSM_unitarity::max_scattering_eigenvalue_infinite_s(m);
+   for (int i = 0; i < size; i++) {
+      for (int j = 0; j < size; j++) {
+         //BOOST_CHECK_CLOSE_FRACTION(unitarityMatrix.second.coeff(i,j), ref_matrix.coeff(i, j), 1e-16);
```
It seems this test does not test anything. Do you want to postpone this?

wkotlarski replied:
I'd leave it for now. I'll make a mark in the FS 3.0 manual that we want to showcase this calculation in the CMSSM and then I'll fix this test to match published example.

Expander replied:
Ok. Please also mark the code in the test with `// @todo(wkotlarski)` (so we can grep for it).

wkotlarski replied:
Good idea. Is this a syntax we should always follow in the code? `// @todo(GITHUB_USERNAME)`?

Expander replied:
Yes, I'd vote for that. :)

Expander commented on test/test_wrappers.cpp:
```diff
@@ -51,6 +51,7 @@ BOOST_AUTO_TEST_CASE( test_Constants )
    BOOST_CHECK_CLOSE_FRACTION(zeta4, std::riemann_zeta(4), 3e-16);
    BOOST_CHECK_EQUAL(zeta5         , std::riemann_zeta(5));
    BOOST_CHECK_EQUAL(ln2           , boost::math::constants::ln_two<double>());
+   BOOST_CHECK_CLOSE_FRACTION(oneOverSqrt2  , 1/sqrt(2.), 3e-16);
```
I'd suggest to write `std::sqrt(2.)` instead of `sqrt(2.)`, because we include `<cmath>`, not `<math.h>`.

wkotlarski replied:
Done in 8b3b474.

wkotlarski commented:
Is there a way to control precision of numbers returned by mathlink interface? Or is it always float?

Expander commented:
I think it is not restricted to single-precision floating point numbers. in `<model>_librarylink.cpp` we use `MTensor` for passing (n-dimensional arrays) numbers. There we check if the elements are of type `MType_Real`, which stands for `double`.

Expander commented:
Please see the following code snippet:
~~~.cpp

DLLEXPORT int FSCMSSMOpenHandle(
   WolframLibraryData libData, mint Argc, MArgument* Args, MArgument Res)
{
   using namespace flexiblesusy;
   using namespace flexiblesusy::CMSSM_librarylink;

   if (Argc != 1)
      return LIBRARY_TYPE_ERROR;

   MTensor pars = MArgument_getMTensor(Args[0]);

   if (libData->MTensor_getType(pars) != MType_Real ||
       libData->MTensor_getRank(pars) != 1)
      return LIBRARY_TYPE_ERROR;

   try {
      auto data = make_data(make_dynamic_array_view(
                          libData->MTensor_getRealData(pars),
                          libData->MTensor_getDimensions(pars)[0]));

      const auto hid = get_new_handle();

      handles.insert(std::make_pair(hid, std::move(data)));

      MArgument_setInteger(Res, hid);
   } catch (const flexiblesusy::Error& e) {
      std::cerr << e.what_detailed() << std::endl;
      return LIBRARY_FUNCTION_ERROR;
   }

   return LIBRARY_NO_ERROR;
}
~~~

wkotlarski commented:
Sorry. My bad. It's actually funny. `make test/test_CMSSM_unitarity.m.xml` prints with a different precision than `wolframscript -f test/test_CMSSM_unitarity.m`. So at least this is fixed.
